src/rust/src/utils/mod.rs

```diff
@@ -743,8 +743,11 @@ pub fn robj_to_rarrow_field(robj: extendr_api::Robj) -> RResult<Robj> {
 
 pub fn robj_to_datatype(robj: extendr_api::Robj) -> RResult<RPolarsDataType> {
     let rv = rdbg(&robj);
-    let res: ExtendrResult<ExternalPtr<RPolarsDataType>> = robj.try_into();
-    let ext_dt = res.bad_val(rv).mistyped(tn::<RPolarsDataType>())?;
+    if rv != "ExternalPtr.set_class([\"RPolarsDataType\"]" {
+        return Err(RPolarsErr::new().bad_val(rv).mistyped(tn::<RPolarsDataType>().to_string()).into());
+    }
```

Comment:
As you said in https://github.com/pola-rs/r-polars/pull/1154#issuecomment-2294958253, checking on this string looks very fragile

Reply:
Oh I see now.. I think this can be addressed in https://github.com/extendr/extendr/pull/853. I just added a test that shows that we can deal with this situation, with very accurate type-checking.